Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+  intro a b hab
+  cases' le_total a 0 with ha ha
+  · cases' le_or_lt b 0 with hb hb
+    · rw [← neg_lt_neg_iff, ← neg_pow_bit1, ← neg_pow_bit1]
+      exact pow_lt_pow_of_lt_left (neg_lt_neg hab) (neg_nonneg.2 hb) (bit1_pos (zero_le n))
+    · exact (pow_bit1_nonpos_iff.2 ha).trans_lt (pow_bit1_pos_iff.2 hb)
+  · exact pow_lt_pow_of_lt_left hab ha (bit1_pos (zero_le n))
+#align strict_mono_pow_bit1 strict_mono_pow_bit1
+
+end bit1
+
+/-- Bernoulli's inequality for `n : ℕ`, `-2 ≤ a`. -/
+theorem one_add_mul_le_pow (H : -2 ≤ a) (n : ℕ) : 1 + (n : R) * a ≤ (1 + a) ^ n :=
+  one_add_mul_le_pow' (mul_self_nonneg _) (mul_self_nonneg _) (neg_le_iff_add_nonneg'.1 H) _
+#align one_add_mul_le_pow one_add_mul_le_pow
+
+/-- Bernoulli's inequality reformulated to estimate `a^n`. -/
+theorem one_add_mul_sub_le_pow (H : -1 ≤ a) (n : ℕ) : 1 + (n : R) * (a - 1) ≤ a ^ n := by
+  have : -2 ≤ a - 1 := by
+    rwa [← one_add_one_eq_two, neg_add, ← sub_eq_add_neg, sub_le_sub_iff_right]
+  simpa only [add_sub_cancel'_right] using one_add_mul_le_pow this n
+#align one_add_mul_sub_le_pow one_add_mul_sub_le_pow
+
+end LinearOrderedRing
+
+namespace Int
+
+#check Nat.cast_pow
+
+/- Porting note: removed `@[simp]` because of linter error:
+
+Left-hand side simplifies from
+  ↑(Int.natAbs x ^ 2)
+to
+  ↑(Int.natAbs x) ^ 2
+using
+  simp only [Nat.cast_pow]
+Try to change the left-hand side to the simplified term!
+
+- Porting note: because of `HMul` more type annotations are needed than in the Lean 3 version.
+-/
+-- @[simp]
+theorem nat_abs_sq (x : ℤ) : ((x.natAbs ^ 2 : ℕ) : ℤ) = x ^ 2 := by rw [sq, Int.natAbs_mul_self, sq]
+#align int.nat_abs_sq Int.nat_abs_sq
+
+alias nat_abs_sq ← nat_abs_pow_two
+
+theorem abs_le_self_sq (a : ℤ) : (Int.natAbs a : ℤ) ≤ a ^ 2 := by
+  rw [← Int.nat_abs_sq a, sq]
+  norm_cast
+  apply Nat.le_mul_self
+#align int.abs_le_self_sq Int.abs_le_self_sq
+
+alias abs_le_self_sq ← abs_le_self_pow_two
+
+theorem le_self_sq (b : ℤ) : b ≤ b ^ 2 :=
+  le_trans le_natAbs (abs_le_self_sq _)
+#align int.le_self_sq Int.le_self_sq
+
+alias le_self_sq ← le_self_pow_two
+
+theorem pow_right_injective {x : ℤ} (h : 1 < x.natAbs) : Function.Injective ((· ^ ·) x : ℕ → ℤ) :=
+  by
+  suffices Function.Injective (natAbs ∘ ((· ^ ·) x : ℕ → ℤ)) by
+    exact Function.Injective.of_comp this
+  convert Nat.pow_right_injective h
+  ext n
+  rw [Function.comp_apply, nat_abs_pow]
+#align int.pow_right_injective Int.pow_right_injective
+
+end Int
+
+variable (M G A)
+
+-- Porting note: dot notation was not working so a roundabout way was used.
+-- Also the components of the structure are given explicitly (maybe possible to avoid this).
+/-- Monoid homomorphisms from `multiplicative ℕ` are defined by the image
+of `multiplicative.of_add 1`. -/
+def powersHom [Monoid M] :
+    M ≃
+      (Multiplicative ℕ →* M) where
+  toFun : M → Multiplicative ℕ →* M :=
+  fun x =>
+    { toFun := fun n : ℕ  =>
+        let n' := Multiplicative.toAdd n
+        x ^ n'
```

Comment:
```suggestion
    { toFun := fun n  =>
        x ^ (Multiplicative.toAdd n)
```
I'm worried that your version leaves a let binding in a definition.

Reply:
```
      map_one' := pow_zero x
      map_mul' := pow_add x
```

Reply:
Thanks. Made changes